llvm/lib/Target/Hexagon/HexagonISelLowering.cpp

```diff
@@ -2156,6 +2157,23 @@ bool HexagonTargetLowering::hasBitTest(SDValue X, SDValue Y) const {
   return X.getValueType().isScalarInteger(); // 'tstbit'
 }
 
+bool HexagonTargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
+
+  if (ShiftLHS->hasOneUse() ||
+      sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add))))))
```

Comment:
what are you trying to achieve here? I think either `ShiftLHS->hasOneUse()` or the sd_match call is redundant

Reply:
This patch remove the restriction for folding (shl (add_nsw x, c1)), c2) and folding (shl(sext(add x, c1)), c2), and test case from dhrystone , see this link:
riscv32: https://godbolt.org/z/o8GdMKrae
riscv64: https://godbolt.org/z/Yh5bPz56z

Reply:
What I meant was that if `ShiftLHS->hasOneUse()` is false, the following sd_match will also be false -- then why do you have both `ShiftLHS->hasOneUse()` _and_ the sd_match call?

Reply:
fixed